### DISCUSSION TITLE
Traveling Wave Solution for Double Obstacle Potential

### DISCUSSION THREAD

**Discussion by Mascsu** [idaholab/moose#19253] [Q&A Modules: Phase field] [answered]
Hello, everyone

I want to use MOOSE to solve the phase-field model with double obstacle potential. The equations are as following:

![FB376AD9-725E-40F3-85A9-59A2B6C66B59](https://user-images.githubusercontent.com/38125871/139358327-bb38c366-a930-400a-a645-e522e5ba5146.png)

The free energy density is defined by Eqs.(64) and (65), the governing equation is Eq.(66) and the parameters are expressed by Eq.(70) where eta, sigma, mu and delta_g are input parameters of materials. (The test values are eta = 6, sigma = mu = 1.0, delta_g = 0.05.)  The reference is from: https://iopscience.iop.org/article/10.1088/0965-0393/17/7/073001/pdf (I. Steinbach, Modell. Phase-field models in materials science. Modelling and Simulation in Materials Science and Engineering. 17(7) (2009) 073001.).

I did not find a suitable built-in kernel to solve this problem. The built-in kernels in phase-field module used double well potential. I have tried to write the custom kernel for double obstacle potential: 

- write a kernel inherited from "ADTimeDerivative" kernel to deal with the left-hand side in Eq.(66);
- write a kernel inherited from "ADKernel" kernel to deal with the first term in the right-hand side;
- write a kernel inherited from "ADKernelValue" kernel to deal with 2nd and 3rd terms in the right-hand side.

It did not work. However, I have used the same way to solve the equations with double well potential successfully. 

Could anyone help me solve this equation?

Thanks,

Sa Ma

**Comment by GiudGiud** [comment]
Hello

is there any way you could solve a simpler equation first, then build from there?
Like what s pretty likely is that there is an issue with one of your kernels, and only by doing this progressively and verifying against known solutions can you know that everything's right. 

Guillaume

**Reply by Mascsu** [reply]
Hello

Thanks for your answer.

I have tried to solve the equation with double well potential and it was successful. But when I used the same way to solve the equation with double obstacle potential, it failed. The only difference between these two equations is the local free energy.  

This is the equation with double well potential:
![image](https://user-images.githubusercontent.com/38125871/139415734-254f384d-2dd2-4bf6-8f3c-9adeb562294c.png)
and this is the equation with double obstacle potential:
![image](https://user-images.githubusercontent.com/38125871/139415791-993e986b-ef0f-4a5b-b75e-636a7672a49d.png)

So, I want to know if any built-in kernel could handle the phase-field model with double obstacle potential. If there is no built-in kernel, could you give me some suggestions how to solve the equations.

Thank you,

Sa Ma

**Reply by GiudGiud** [reply]
Ok I dont know the answer to that so we ll ping @laagesen. 

Can you try turning sqrt(phi ( 1 - phi) ) into sqrt(phi ( 1 - phi) + 1e-16) ? The singularity of the square root could be messing with AD

**Reply by Mascsu** [reply]
Thanks for your answer,. I have tried turning sqrt(phi ( 1 - phi) ) into sqrt(phi ( 1 - phi) + 1e-16), but it did not work.

**Comment by laagesen** [comment, marked as answer]
Hi Sa Ma - the double obstacle potential is difficult to implement using any code that uses an implicit solution method, including MOOSE. The reason is that most implementation use a free energy that shoots up to infinity outside the range 0 < phi < 1. This huge discontinuity in the free energy makes it very difficult to solve with an implicit method. As far as I know, all codes that use this formulation are explicit codes. MOOSE does have the capability to solve time steps explicitly, but to be totally honest it is not that well suited to phase-field models with the double obstacle potential. That is one of the reasons why we don't have any kernels/materials in the framework for this problem.

What is the physical problem you are interested in? We have alternative formulations that may be able to accomplish what you want to do already implemented.

**Reply by GiudGiud** [reply]
There's an explicit time integrator in MOOSE. 
https://mooseframework.inl.gov/source/timeintegrators/ActuallyExplicitEuler.html
but maybe it doesnt work with phase field?

**Reply by laagesen** [reply]
It probably does work, but no one has tried it for phase-field that I know of. The problem is that the stability limit for explicit time steps will prevent you from taking large time steps, so the problem will run slowly. Most phase-field problems are much better suited to implicit time integration schemes, but the double obstacle potential is an exception. 

If others have tried the explicit time integrator on phase-field problems, please chime in.

**Reply by Mascsu** [reply]
Hello, thanks for your answer!

I want to solve the phase-field model with finite interface dissipation. Different from kks model, this model could deal with the non-equilibrium phase transition. In the model, the double obstacle potential is used. Here is the website of the paper: https://www.sciencedirect.com/science/article/pii/S1359645412000730.

Maybe I have two ways to deal with the problem. 

(I) I can replace the origin double obstacle potential with the double well poential and re-derive the phase-field model. But I am not sure whether the implicit algorithm could successfully solve the compostion field equation, which includes strongly non-linear terms. The equations are as following
![image](https://user-images.githubusercontent.com/38125871/139639632-a0cf1b9b-abb0-4ec9-a261-1ebd726b31d5.png)

(II) I can try the explicit time integrator. I have used the finite difference method (FDM) to solve the phase-field model with double obstacle potential. I need to cutoff the unphysical value (if phi < 0, make phi = 0; else if phi >1, make phi = 1) when I use the FDM. I am not sure how to make sure the value of phi always between 0 and 1 in MOOOSE framework.

Thanks,

Sa Ma

**Reply by GiudGiud** [reply]
Hello

For (II) you can use the Bounds or the Dampers system to keep the value between 0 and 1
https://mooseframework.inl.gov/source/auxkernels/ConstantBoundsAux.html
https://mooseframework.inl.gov/source/dampers/BoundingValueElementDamper.html

Guillaume

**Reply by laagesen** [reply]
I'd recommend starting with the Bounds system, some people have used this for phase-field fracture modeling with success. Please let us know how it goes! You may not need explicit time stepping if you can use Bounds successfully.

**Reply by Mascsu** [reply]
Thanks for your answers!

I have used the Bounds system (ConstantBoundsAux) and implicit time stepping, and the results agree well with the analytical solution.

**Reply by laagesen** [reply]
Very interesting, and thanks for letting us know! Tagging @dschwen FYI

**Reply by dschwen** [reply]
Good to know. It would be really cool if you could contribute your code back. It would make a great example for the use of the double obstacle potential together with the bounds system. We can help you get everything in shape if you make a pull request.

**Reply by Mascsu** [reply]
I have made a pull request. Next, I will try to solve the phase-field model with finite interface dissipation.
https://www.sciencedirect.com/science/article/pii/S1359645412000730.
